File: code/frontend/src/api/prediction.ts
// ============================================================================
// Prediction API - 能耗预测
// ----------------------------------------------------------------------------
// 3 个 endpoint, 全部挂在 /api/v1 前缀下:
//   POST /prediction/buildings/{id}/forecast  提交预测 job (202, demo 也可用)
//   GET  /prediction/jobs/{id}                 查 job 详情 (前端轮询到 SUCCEEDED/FAILED)
//   GET  /prediction/buildings/{id}/latest     取最近一次 SUCCEEDED 预测 (进面板时先调这个)
//
// 异步流程: POST 拿 job_id -> 每 2s 轮询 GET /jobs/{id} -> SUCCEEDED 拿 result_jsonb
// result_jsonb 结构: { history, forecast, mape, model_type, warning }
//   history:  训练用过的历史日数据 [{date, value}]
//   forecast: 未来 horizon 天预测 [{date, yhat, yhat_lower, yhat_upper}]
//   mape:     最后 eval_days 天实际 vs 预测算的 MAPE (越小越好, <20% 优秀)
//   warning:  训练数据不足提示 (BDG2 demo 单年数据 Prophet 学不到年季节性)
// ============================================================================

import { api } from './client'

// ---------------------------------------------------------------------------
// 模型类型 - 对齐后端 mart.prediction_job.model_type CHECK 约束
//   prophet: Facebook 时序预测, 主力 (年/周/趋势分量自动学习)
//   lstm:    PyTorch LSTM demo (CPU, 50 epoch, 自回归预测)
//   linear:  sklearn LinearRegression baseline (X=day_index, 给上面两个做参考)
//   gbm:     sklearn GradientBoosting 天气驱动回归 (气温/露点/风速/云量 + 时间特征)
// ---------------------------------------------------------------------------
export type PredictionModel = 'prophet' | 'lstm' | 'linear' | 'gbm'

// gbm 特征重要性 [{feature, importance}], importance 0-1 和=1, 降序
export interface FeatureImportance {
  feature: string
  importance: number
}

// job 状态机: PENDING -> RUNNING -> SUCCEEDED / FAILED
// 失败立即置 FAILED 不自动重试 (与 triposplat 不同, 详见后端 prediction_worker)
export type PredictionJobStatus = 'PENDING' | 'RUNNING' | 'SUCCEEDED' | 'FAILED'

// ---------------------------------------------------------------------------
// result_jsonb 里的子结构
// ---------------------------------------------------------------------------
export interface PredictionHistoryPoint {
  date: string         // YYYY-MM-DD
  value: number        // kWh, 当天总能耗
}

export interface PredictionForecastPoint {
  date: string                 // YYYY-MM-DD
  yhat: number                 // 预测值
  yhat_lower: number | null    // 置信区间下界 (Prophet 给, LSTM/linear 给 null)
  yhat_upper: number | null    // 置信区间上界
}

export interface PredictionResult {
  history: PredictionHistoryPoint[]
  forecast: PredictionForecastPoint[]
  mape: number | null          // 百分比, 例如 6.21 表示 6.21%
  model_type: PredictionModel
  warning: string | null       // 训练数据不足提示, null 表示无警告
  // gbm 特征重要性 (只有 gbm 模型有, 其他模型 undefined)
  feature_importances?: FeatureImportance[]
}

// ---------------------------------------------------------------------------
// job 详情 (GET /prediction/jobs/{id} 和 /latest 共用)
//   status=SUCCEEDED 时 result 字段有值
//   status=FAILED    时 error_message 有值
//   status=PENDING/RUNNING 时 result/error_message 都为 null
// ---------------------------------------------------------------------------
export interface PredictionJob {
  id: string
  building_id: string
  model_type: PredictionModel
  horizon_days: number
  status: PredictionJobStatus
  mape: number | null
  result: PredictionResult | null
  error_message: string | null
  started_at: string | null
  finished_at: string | null
  created_at: string
}

// ---------------------------------------------------------------------------
// 请求体 - POST /prediction/buildings/{id}/forecast
//   horizon_days: 预测天数 1-90 (后端 CHECK 约束)
//   model_type:   模型类型
// ---------------------------------------------------------------------------
export interface ForecastRequest {
  horizon_days: number
  model_type: PredictionModel
}

export const predictionApi = {
  // 提交预测 job. 202 Accepted + job_id. demo 也可用 (异步只读任务, 不写 fact 表).
  createForecast(buildingId: string, body: ForecastRequest) {
    return api.post<{ job_id: string; status: 'PENDING' }>(
      `/prediction/buildings/${buildingId}/forecast`,
      body,
    )
  },

  // 查 job 详情. 前端轮询直到 status=SUCCEEDED 或 FAILED.
  getJob(jobId: string) {
    return api.get<PredictionJob>(`/prediction/jobs/${jobId}`)
  },

  // 取最近一次 SUCCEEDED 的预测. 进 PredictionPanel 时先调这个, 有缓存直接显示.
  // 没跑过预测返 200 + null (后端把"没跑过"当合法初始态, 不返 404).
  getLatest(buildingId: string, params?: { model_type?: PredictionModel }) {
    return api.get<PredictionJob | null>(`/prediction/buildings/${buildingId}/latest`, { params })
  },
}

File: code/backend/app/api/prediction.py
"""
能耗预测路由。

三个接口:
  POST /prediction/buildings/{id}/forecast   提交预测 job (202, demo 例外可调)
  GET  /prediction/jobs/{id}                  查 job 详情 (读)
  GET  /prediction/buildings/{id}/latest      取最近一次 SUCCEEDED 预测 (读)

路由薄壳: 业务在 prediction_service。路由层只做:
  - building 存在性校验 (404)
  - horizon_days / model_type 参数校验 (Pydantic 在请求体层就做了)
  - job 不存在/跨租户转 404
  - 响应包装 (success/error)

submit_forecast 挂 get_current_user 而非 require_write_access: 预测是异步只读
任务, 不写 fact 表只写 mart.prediction_job (带 tenant_id 隔离), demo 也能体验.
其他写接口 (上传/重建/回滚等) 仍挂 require_write_access 拦 demo.
"""
from fastapi import APIRouter, Depends, HTTPException, Query, status
from loguru import logger

from app.core.deps import CurrentUser, get_current_user
from app.core.response import error, success
from app.db.session import get_conn
from app.models.prediction import ForecastRequest
from app.services.prediction_service import (
    create_job,
    get_job,
    get_latest,
)


router = APIRouter(tags=["prediction"])


def _verify_building_access(building_id: str, tenant_id: str) -> None:
    """校验 building 属于当前租户, 不存在抛 404。

    跟 api/reconstruction.py 的同名函数一致, 一期没抽公共模块
    (后续如果跨模块重复变多再考虑抽 app/api/_common.py)。
    """
    with get_conn() as conn:
        with conn.cursor() as cur:
            cur.execute("""
                SELECT id FROM core.building
                WHERE id = %s::uuid AND tenant_id = %s::uuid
            """, (building_id, tenant_id))
            if cur.fetchone() is None:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=error(message=f"building 不存在或租户越权: {building_id}", code=404),
                )


@router.post(
    "/prediction/buildings/{building_id}/forecast",
    status_code=status.HTTP_202_ACCEPTED,
)
def submit_forecast(
    building_id: str,
    req: ForecastRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """提交预测 job。202 Accepted (异步任务, prediction_worker 进程消费)。

    body:
      horizon_days: 1-90
      model_type: prophet / lstm / linear

    worker 启动方式 (单独进程):
      cd backend
      python -m worker.prediction_worker

    前端拿到 job_id 后轮询 GET /prediction/jobs/{id} 直到 SUCCEEDED/FAILED。

    demo 例外: 这里挂 get_current_user 而不是 require_write_access, 让 demo
    账号也能体验预测. 理由: 预测是异步只读任务, 不写 fact 表只写
    mart.prediction_job (该表带 tenant_id 隔离, demo 跑的 job 只属于 demo
    租户, 不会污染其他租户). 其他写接口 (上传/重建/回滚等) 仍挂
    require_write_access 拦 demo, 那些会改 fact 数据或文件系统.
    """
    _verify_building_access(building_id, user.tenant_id)

    job_id = create_job(
        tenant_id=user.tenant_id,
        building_id=building_id,
        model_type=req.model_type,
        horizon_days=req.horizon_days,
    )

    logger.info(
        "submit_forecast tenant={} building={} job={} model={} horizon={}",
        user.tenant_id[:8], building_id[:8], job_id[:8],
        req.model_type, req.horizon_days,
    )
    return success(
        data={"job_id": job_id, "status": "PENDING"},
        message="预测任务已提交, 等待 worker 消费",
    )


@router.get("/prediction/jobs/{job_id}")
def get_job_status(
    job_id: str,
    user: CurrentUser = Depends(get_current_user),
):
    """查 job 状态。前端轮询直到 SUCCEEDED/FAILED。

    返回 result_jsonb (含 history + forecast + mape + warning)。
    """
    result = get_job(user.tenant_id, job_id)
    if result is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=error(message=f"job 不存在或租户越权: {job_id}", code=404),
        )
    return success(data=result)


@router.get("/prediction/buildings/{building_id}/latest")
def get_latest_forecast(
    building_id: str,
    model_type: str | None = Query(
        None,
        pattern="^(prophet|lstm|linear|gbm)$",
        description="按模型类型过滤, 不传则取任意模型的最近预测",
    ),
    user: CurrentUser = Depends(get_current_user),
):
    """取最近一次 SUCCEEDED 的预测结果。model_type 可选过滤。

    前端默认进入 PredictionPanel 时调这个, 有最近预测直接显示避免重新跑。
    没有则返 200 + data=null (前端显示空状态, 提示用户点预测按钮)。

    "没跑过预测"是合法初始态而非"资源不存在", 返 200 + null 让前端按空状态处理,
    避免浏览器 Network 面板对 4xx 默认打红字干扰调试。跟 visual-model 接口
    设计对齐: 没设过 visual_model 也返 200 + null。
    """
    _verify_building_access(building_id, user.tenant_id)

    result = get_latest(
        tenant_id=user.tenant_id,
        building_id=building_id,
        model_type=model_type,
    )
    return success(data=result)
